Map iterator

cwlittle opened the pull request.

cwlittle committed:
Add Next trait and primitive implementations

cwlittle committed:
Refactor impl Next blocks for macro

cwlittle committed:
Add Map Iterator architecture with placeholder next function

cwlittle committed:
Add MapIterator new method

cwlittle committed:
Update get_start_key method doc string

cwlittle committed:
Move auxilliary test function

cwlittle committed:
Add get_next on map::Map

map.children change from HashMap -> BTreeMap to enable .range() method
to be able to be called. This had a few ramifications where additional
trait bounds were required for the entirety of the Map, like Ord.

cwlittle committed:
Cargo fmt

cwlittle committed:
Remove superfluous Deref bounds

cwlittle committed:
Add test for get_next in-memory map case

cwlittle committed:
Add test for get_next with no entry key

cwlittle committed:
Add test for get_next on empty map

cwlittle committed:
Update get_next to always return the key with None if no key exists

cwlittle committed:
Remove get_next in favor of iterator merge

cwlittle committed:
Remove get_next tests

cwlittle committed:
Add initial iter_merge_next implementation

cwlittle committed:
Remove superflous imports

cwlittle committed:
Add test for map only iter_merge_next

cwlittle committed:
Add test for store only iter_merge_next

cwlittle committed:
Add test for removal from map before flush

cwlittle committed:
Add test for in-memory map out of store range

Calling peekable on iterators at every iteration caused peek to consume
the first value of the iterator. Changed the function to take Peekable<>
instead of iterators.

cwlittle committed:
Update tests to account for peekable change

cwlittle committed:
Add test for next merge iter store value in range of map

cwlittle committed:
Add iter merge next test for map empty key

cwlittle committed:
Add iter merge next test for map empty key store non empty

cwlittle committed:
Update MapIterator implementation

cwlittle committed:
Add rudimentary iter implementation to map

cwlittle committed:
Add test for iter in memory map only

cwlittle committed:
Add test for

cwlittle committed:
Add test for iter map entry removal

cwlittle committed:
Add test for map iter map key out of store range

cwlittle committed:
Add test for iter store key in map range

cwlittle committed:
Add test for iter map entry deletion

cwlittle committed:
Add test for iter map entry deletion store non-empty

cwlittle committed:
Clean up iter function

cwlittle committed:
Add rudimentary range function

cwlittle committed:
Add rudimentary range implementation

cwlittle committed:
Add test for map range unbounded

cwlittle committed:
Add test for map range bounded start

cwlittle committed:
Add test for map range end bounded non-inclusive

cwlittle committed:
Add test for map range end bounded inclusive

cwlittle committed:
Add test for map range bounded non-inclusive

cwlittle committed:
Add test for map range bounded inclusive

cwlittle committed:
Change iter implementation to call map

cwlittle committed:
Merge branch 'query' into map-iterator

cwlittle committed:
Clippy fix: Remove redundant clone

cwlittle committed:
Clippy fix: needless return

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: unused associated function

cwlittle committed:
Clippy fix: unused variable

cwlittle committed:
Clippy fix: unused struct field

cwlittle committed:
Clippy fix: clone on copy

cwlittle committed:
Clippy fix: suppress unused imports in test mod

cwlittle committed:
Clippy fix: dead code

cwlittle committed:
Clippy fix: dead code

cwlittle committed:
Cargo fmt

cwlittle commented:
No CI triggered for merge into query branch. Clippy and fmt good on my machine. Anything that lingers can be addressed when query merges into develop.

cwlittle committed:
Add test for map iter store only bounded

cwlittle committed:
Add test for map range bounded

cwlittle committed:
Add test for map range empty

cwlittle committed:
Remove new lines

cwlittle committed:
Add test for map of map

cwlittle committed:
Add test map of map from store

cwlittle committed:
Add test for map of deque

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+    K: Encode + Decode + Terminated + Eq + Hash + Next<K> + Ord + Copy,
+    V: State<S> + Decode + Copy,
+    S: Read,
+{
```
This feels like it shouldn't be an associated function on `Map` - maybe it should be a standalone function or on `MapIterator` instead?

cwlittle replied:
Moved this as an associated function on MapIterator

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+impl<'a, 'b, K, V, S> Map<K, V, S>
+where
+    K: Encode + Decode + Terminated + Eq + Hash + Next<K> + Ord + Copy,
+    V: State<S> + Decode + Copy,
```
This bound isn't quite right - if we require that `V` has `Decode` then we won't be able to support complex types as `V`. The `State` trait has the associated type `Encoding` which is what defines the `Encode + Decode` implementations, so to support complex types you should decode values by decoding them into `<V as State<S>>::Encoding`, then turning it into `V` with `State<S>::create()` (which will also involve creating a `Store` with the desired key prefix based on the encoded `K`).

mappum replied:
Oh, this also shouldn't require `Copy` since that also constrains what we can iterate over, although I assume if you needed this then there might be some things in the current implementation that make it hard to remove. 🤔

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+
+                // consumed map iterator, still have backing values
+                (false, true) => {
+                    match store_iter.next().transpose()? {
```
I think this match block could be cleaner if it was just an `.expect(...)`, but feel free to ignore if you like this style better.

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+
+                    // map_key > backing_key, emit the backing entry
+                    // map_key == backing_key, map entry shadows backing entry
+                    if key_cmp == Ordering::Greater || key_cmp == Ordering::Equal {
```
There's an issue here, we don't want to emit the backing entry if the keys are equal (we should be advancing the backing iterator and dropping what it returns, then returning the entry emitted by the map iterator).

There would currently be a bug if we iterated over an updated value, the old value would be emitted instead.

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+                    if key_cmp == Ordering::Greater || key_cmp == Ordering::Equal {
+                        let entry = store_iter.next().unwrap()?;
+                        let decoded_key: K = Decode::decode(entry.0.as_slice())?;
+                        let encoded_key: V = Decode::decode(entry.1.as_slice())?;
```
As I commented above, this should decode into `V::Encoding` then construct `V` using `V::create()`. Also, this should be called something like `decoded_value` rather than `encoded_key`.

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+        }
+    }
+
+    pub fn iter(&'a mut self) -> MapIterator<'a, K, V, S> {
```
This shouldn't need a mutable reference, if we want a mutable version we could add an `iter_mut`.

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +330,32 @@ where
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match Map::iter_merge_next(&mut self.map_iter, &mut self.store_iter) {
+            Err(err) => panic!("{}", err),
```
We don't want this to be able to panic otherwise nodes will possibly be able to crash. Instead, `Item` should be wrapped in `Result`.

mappum commented on src/collections/map.rs:
```diff
@@ -163,6 +166,114 @@ where
+    K: Encode,
+{
+    match bound {
+        Bound::Included(inner) => Bound::Included(Encode::encode(inner).unwrap()),
```
This shouldn't `unwrap` since it can possibly cause a panic, the return value should be wrapped in a `Result` instead.

mappum replied:
On second thought, I'm not sure any encode implementations that exist now can possibly panic, so this could be OK.

We could leave this and create a marker trait in `ed` for encodings to specify that they cannot fail, or we could just add the ugly `Result`-wrapping to ensure safety in the meantime. 🤔

mappum commented on src/collections/map.rs:
```diff
@@ -281,12 +418,12 @@ pub enum ChildMut<'a, K, V, S> {
 impl<'a, K, V, S> ChildMut<'a, K, V, S>
 where
-    K: Hash + Eq + Encode + Terminated,
+    K: Hash + Eq + Encode + Terminated + Ord + Copy,
```
We shouldn't need to add bounds here if the implementation didn't change - they don't have to be the same across all impl blocks (it's OK to have methods which are defined for some types of `K` but not others).

In general, I try to use the minimum number of bounds required for each scope (sometimes moving methods into their own impl block, but it's OK to not follow this rule religiously if in practice it won't really make the method apply to more types).

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +330,32 @@ where
     }
 }
 
+pub struct MapIterator<'a, K, V, S>
```
To make this more idiomatic, it could just be called `Iter` (it's already implied that it's a map iterator since the full path is `orga::collections::map::Iter`).

mappum commented on src/collections/map.rs:
```diff
@@ -320,17 +457,25 @@ impl<'a, K, V, S> Deref for ChildMut<'a, K, V, S> {
-                let entry = parent.children.entry(key).insert(Some(value));
-                *self = ChildMut::Modified(entry);
-                self.deref_mut()
+                let _insertion = parent.children.insert(key, Some(value));
```
Just a small optimization, but this can avoid doing 2 traversals through the btree by inserting inside the match arms, e.g. something like:
```rust
btree_map::Entry::Occupied(entry) => {
    entry.insert(Some(value));
    *self = ChildMut::Modified(entry);
    self.deref_mut()
}
// and similar for Vacant
```

mappum commented on src/collections/map.rs:
```diff
@@ -562,7 +712,650 @@ mod tests {
+                assert_eq!(*value, 24);
+            }
+            None => assert!(false),
+        }
```
I would expect the removal to apply to the values emitted by the iterator, the test should ensure the entry isn't emitted.

mappum commented on src/collections/map.rs:
```diff
@@ -562,7 +712,650 @@ mod tests {
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(*actual, 84);
```
Maybe I missed it, but I don't think there's a test which covers updating a value which is already in the store, then trying to iterate over it (which would have caught an issue in `iter_merge_next`).

cwlittle committed:
Move Map::iter_get_next -> MapIterator::iter_get_next

cwlittle committed:
Swap match statement for .expect()

cwlittle committed:
Fix bug in iter_merge_next

Split the logic of Ordering::Equal to pass over values rather than
returning the old value if it has been updated since the store was
reinstated

cwlittle committed:
Remove mutable requirement for iter and range

cwlittle committed:
Remove panic in MapIterator::next()

Changed MapIterator associated type Item to Result<(K, Child<'a, V)> to
avoid panic

cwlittle committed:
Replace .unwrap with Result in encode_bounds

This causes both iter and range to return Result<MapIterator>> requiring
that both are unwrapped in tests and ?'d in user functions

cwlittle committed:
Remove Copy trait from MapIterator

cwlittle committed:
Rename Child to Ref

cwlittle committed:
Rename Ref Unmodified variant to Owned

cwlittle committed:
Rename Ref Modified variant to Borrowed

cwlittle committed:
Update MapIterator to return Ref<K> in place of K

cwlittle committed:
Update tests to account for returned key reference

cwlittle committed:
Update deque for Child -> Ref

cwlittle committed:
Update deref_mut method

Because btree_map.entry() does not have an insert method, there is no
way to both insert a value and get the entry from the key without the
requirement of reuse of the key. Therefore, the Copy trait is required
to insert a value and get the corresponding OccupiedEntry that is
required for deref_mut

cwlittle committed:
Remove Decode trait from V in MapIterator

cwlittle committed:
Add Clone trait remnants

cwlittle committed:
Alias store::Iter as StoreIter

cwlittle committed:
Rename MapIterator -> Iter

cwlittle committed:
Rename used store::Iter -> StoreIter

cwlittle committed:
Add test for iterating over overwritten store value

This contains map.insert() which is not yet implemented

cwlittle committed:
Add map insert method

cwlittle committed:
Update insert method to account for complex types

cwlittle committed:
Add test for normal map insertion

cwlittle committed:
Add test for map insertion with complex types

cwlittle committed:
Add test for map insertion with flush

cwlittle committed:
Add test for map insert overwrite store entry

cwlittle committed:
Add test for map insert store overwrite entry valid insert

cwlittle committed:
Add unwrap to map insert in test

cwlittle committed:
Rollback closure arg naming in Ref

mappum commented on src/collections/map.rs:
```diff
@@ -44,7 +49,7 @@ where
         S: Write,
     {
-        for (key, maybe_value) in self.children.drain() {
+        for (key, maybe_value) in IntoIterator::into_iter(self.children) {
```
Do we need to explicitly call `into_iter` here? For loop syntax should already call it for what you pass in, so I think this can be `for (key, maybe_value) in self.children`.

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +273,123 @@ where
+    V: State<S>,
+    S: Read,
+{
+    fn iter_merge_next(
```
Now that this takes in all the fields of `Iter` as parameters, it'd probably be a little simpler by just being a method that just takes a `&mut Iter`.

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +273,123 @@ where
+                    let entry = store_iter
+                        .next()
+                        .transpose()?
+                        .expect("Peek ensures that this in unreachable");
```
in -> is

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +273,123 @@ where
+                    let key_cmp = map_key.cmp(&decoded_backing_key);
+
+                    // map_key > backing_key, emit the backing entry
+                    // map_key == backing_key, map entry shadows backing entry
```
Let's move this comment line down to the next block

mappum commented on src/collections/map.rs:
```diff
@@ -219,6 +273,123 @@ where
+    type Item = Result<(Ref<'a, K>, Ref<'a, V>)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Iter::iter_merge_next(self.parent_store, &mut self.map_iter, &mut self.store_iter)
```
I mentioned that `iter_merge_next` could be a method, I guess it could also just all be here inside `next`, not sure if one way is cleaner.

cwlittle replied:
Good point. Now that the iter_merge_next takes an Iter, it doesn't make sense to have item_merge_next at all when self can be referenced directly from the next method. In any case, item_merge_next is already a method on Iter, so I think it's a bit cleaner to have next contain all of the next logic rather than calling item_merge_next and having a separate implementation block.

cwlittle replied:
Scratch that. iter_merge_next returns Result<Option<(_, _)>> and next wants that transposed. The single transpose on the returned value from iter_merge_next will be a bit cleaner than the refactored alternative in the next logic.

mappum commented on src/collections/map.rs:
```diff
@@ -245,29 +416,29 @@ impl<V> ReadOnly<V> {
 }
 
 /// An immutable reference to an existing value in a collection.
```
I guess this should now say "an existing **key or** value"

mappum reviewed (changes requested):
Just commented a few small final nitpicks.

Seems sound and I appreciate the thoroughness of the unit tests. :+1:

cwlittle committed:
Remove explicit IntoIterator call

cwlittle committed:
Update iter_merge_next to take Iter struct

cwlittle committed:
Fix typo in match arm

cwlittle committed:
Move block comment to appropriate block

cwlittle committed:
Remove superfluous trait bound from Iterator impl for Iter

cwlittle committed:
Update Ref doc comment
